Zephyr module variables in Kconfig

tejlmand opened the pull request:
With the introduction of https://github.com/zephyrproject-rtos/zephyr/pull/26715 it became apparent that downstream users of Zephyr were dependent on `SOC_DIR` variable in Kconfig.

Because of this, and to align Kconfig and CMake Zephyr module variables between Kconfig and CMake, then this PR proposes to export `ZEPHYR_<module>_MODULE_DIR` to Kconfig.

In CMake today, if module `FOO` is part of Zephyr, then the following is possible in CMake:
```
include(${ZEPHYR_FOO_MODULE_DIR}/myfile.cmake)
```

with this PR, the similar is now possible in Kconfig:
```
source "$(ZEPHYR_FOO_MODULE_DIR)/Kconfig.myfile"
```

This means that users who previously did:
```
source "$(SOC_DIR)/$(ARCH)/Kconfig.mysoc"
```

can now do:
```
source "$(ZEPHYR_MYSOC_MODULE_DIR)/soc/$(ARCH)/Kconfig.mysoc"
```
if the repo `MYSOC` is also a Zephyr module.

--------

Edit: Based on comments, then additional commits has been added that does general clean-up of CMake/python code related to Zephyr modules.

tejlmand committed:
cmake: zephyr modules variable names.

This commit refactors processing of Zephyr modules.
The Zephyr module name processing has been moved to zephyr_modules.cmake
to ensure that all `ZEPHYR_<MODULE_NAME>_MODULE_DIR` are defined early
in the build system.

No changes to inclusion of Zephyr module CMake code.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
@dkalowsk FYI

dkalowsk commented on doc/guides/modules.rst:
```diff
@@ -75,6 +75,25 @@ module:
+When a module has been described using the :file:`module.yml` file then it will
+automatically be included into the Zephyr Kconfig and CMake build system.
+
+To facilitate referencing between Zephyr and Zephyr modules, then all modules
```
minor edit suggestion, drop the `then`.

tejlmand replied:
fixed

mbolivar-nordic commented on cmake/kconfig.cmake:
```diff
@@ -63,6 +63,20 @@ set(ENV{ARCH_DIR}   ${ARCH_DIR})
+       ZEPHYR_KCONFIG_MODULES_DIR
+       "ZEPHYR_${MODULE_NAME_UPPER}_MODULE_DIR=${ZEPHYR_${MODULE_NAME_UPPER}_MODULE_DIR}"
+  )
+  set(ENV{ZEPHYR_${MODULE_NAME_UPPER}_MODULE_DIR}
+      ${ZEPHYR_${MODULE_NAME_UPPER}_MODULE_DIR}
+  )
```
Why is this necessary to set in the main build environment? I would guess adding it to the environments of each kconfig_target would be enough. We have too many environment variables already IMO.

mbolivar-nordic replied:
(This one is not a blocker, just asking if we can avoid adding extra environment variables to the cmake process.)

tejlmand replied:
I completely agree, however this commit follows the existing lines here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/kconfig.cmake#L51-L64
and is currently needed because an `execute_process()` here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/kconfig.cmake#L213-L216

Cleaning up this is outside scope of this PR, but con follow in a dedicated PR covering all above mentioned lines.

mbolivar-nordic replied:
> and is currently needed because an `execute_process()` here:
> https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/kconfig.cmake#L213-L216

Got it. Thanks for the reference.

tejlmand replied:
included a cleanup commit.

mbolivar-nordic commented on doc/guides/modules.rst:
```diff
@@ -75,6 +75,25 @@ module:
+To facilitate referencing between Zephyr and Zephyr modules, then all modules
+will be accessible through Kconfig and CMake variables.
+
+In Kconfig and CMake, the ``ZEPHYR_<module-name>_MODULE_DIR`` can be used to
```
I didn't raise a bug about this because it wasn't officially documented before, but this is not really correct.

The existence of a module.yml file does not guarantee this variable is defined, even in CMake.

See this source code comment in one of my PRs for details: https://github.com/zephyrproject-rtos/zephyr/pull/27688/files#diff-a1f9e35746e0274b9ad8cfb12162414dR38

I think this situation should be fixed, or at least documented more clearly.

tejlmand replied:
Agreed, as minimum this should be rephrased to be more precise that only if build files are present in the module.

Guess the introduction of sanity in `module.yml` made a small hole here.

mbolivar-nordic replied:
> as minimum this should be rephrased to be more precise that only if build files are present in the module.

I agree that would be the minimum acceptable change, since it makes the documentation correct.

mbolivar-nordic replied:
(Although FTR I would appreciate it if the existence of a module.yml resulted in the definition of ZEPHYR_FOO_MODULE_DIR. That is the best possible world but I agree it's outside the scope of this PR)

tejlmand replied:
couldn't leave code like that.
Fixed. Now docs and code is aligned.

mbolivar-nordic replied:
OK, rebased and tested on top of the updated version. I confirm the fix. Thanks for fixing :+1:

dkalowsk commented:
I don't have the ability to fully test out the PR within my current setup as my SOC_DIR is not a module, but making some quick changes the proposed work flow does appear to work.  AKA I can boot on a UUT with the changes.  

My only real comment on this change is that the PR cover documentation.  I don't know what the documentation process is like for Zephyr so feel free to ignore this.  The PR cover documentation gives a little more detail than what the proposed included documentation changes have.  As someone new coming to setup, it would be really useful to have those details included in the documentation, and the reason for why this method is better than another method.

tejlmand commented:
> As someone new coming to setup, it would be really useful to have those details included in the documentation, and the reason for why this method is better than another method.

That's a good point. I will try to add more detailed example, but I will try to avoid referring `SOC_DIR` in the docs.
Reason for that, is that new comers to Zephyr should not be confused about `SOC_DIR` as it is no longer valid.
(Such info belongs to the release notes.)

But for reviewers of this PR, I believe it is of importance to know the motivation for this enhancement, and there the `SOC_DIR` example is of importance.

tejlmand committed:
cmake: kconfig cmake file environment cleanup

This commit is a cleanup of the cmake/kconfig.cmake file.
It removes `set(ENV{<name>} <value>)` and instead creates a shared list
that can be used for execute_process() and add_custom_target() when
running Kconfig during CMake configure time or menuconfig and related
build targets.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: zephyr module processing improvements.

This commit improves Zephyr modules support in CMake.
All Zephyr modules will now have a corresponding
`ZEPHYR_${module_name}_MODULE_DIR` set, regardless of whether they
define a cmake folder in module.yml.

This results in:
`ZEPHYR_${module_name}_MODULE_DIR` defined for all modules and refers to
the root of the modle.
`ZEPHYR_${module_name}_CMAKE_DIR` defined for modules that specifies
cmake in module.yml, or is having both a CMakeLists.txt and Kconfig file
in the zephyr folder.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
scripts: zephyr_modules.py cleanup.

This commit has no functional changes but is simply a cleanup so that
flake8 will pass without warnings.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: export Zephyr modules as Kconfig variables.

In CMake it is possible to reference each Zephyr module as a CMake
variable. This commit introduces the same feature to Kconfig, so that
it is possible to reference a module in consistent way in both CMake
and Kconfig.

To include a file from module FOO in CMake, use:
include(${ZEPHYR_FOO_MODULE_DIR}/somefile.cmake

To source a file from module FOO in Kconfig, use:
source "$(ZEPHYR_FOO_MODULE_DIR)/Kconfig.somefile"

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
docs: Zephyr module and CMake processing

This commit describes add a section which describes how Zephyr modules
are integrated into Zephyr build system and what Zephyr module variables
that are available during CMake and Kconfig processing.

Those variables may be used for refering other Zephyr modules or
including of additional code.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> As someone new coming to setup, it would be really useful to have those details included in the documentation, and the reason for why this method is better than another method.

@ dkalowsk Done, example updated.
Feedback is welcome.

mbolivar-nordic commented on doc/guides/modules.rst:
```diff
@@ -75,6 +75,65 @@ module:
+- Module directory in Kconfig: ``$(ZEPHYR_FOO_MODULE_DIR)``
+- :file:`CMakeLists.txt` directory include in build system: ``${ZEPHYR_FOO_CMAKE_DIR}``
+
+Those variables may be used to obtain a full path to Zephyr module or to test
+whether a given module is present in the system.
+
+For example, to verify that ``FOO`` exists in the system and is a Zephyr module
+then one may do:
```
```suggestion
Notice how a lowercase module name ``foo`` is capitalized to ``FOO``
in both CMake and Kconfig.

These variables can also be used to test whether a given module exists.
For example, to verify that ``foo`` is the name of a Zephyr module:
```

mbolivar-nordic commented on doc/guides/modules.rst:
```diff
@@ -75,6 +75,65 @@ module:
+    # Do something if FOO exists.
+  endif()
+
+In Kconfig, the variable may be used to specify a path to include.
+An example can be inclusion of additional Kconfig SoC files.
+
+In this example ``FOO`` specifies an additional ``SOC_ROOT`` and in
+:file:`<FOO>/soc/arm/foo_soc/Kconfig.soc` one may do:
```
```suggestion
In Kconfig, the variable may be used to find additional files to include.
For example, to include the file :file:`some/Kconfig` in module ``foo``:
```

mbolivar-nordic commented on doc/guides/modules.rst:
```diff
@@ -75,6 +75,65 @@ module:
+
+.. code-block:: kconfig
+
+  source "$(ZEPHYR_FOO_MODULE_DIR)/soc/arm/foo_soc/Kconfig.soc.common"
```
```suggestion
  source "$(ZEPHYR_FOO_MODULE_DIR)/some/Kconfig"
```

mbolivar-nordic commented on doc/guides/modules.rst:
```diff
@@ -75,6 +75,65 @@ module:
+During CMake processing of each Zephyr module, the following two variables are
+also available:
+
+- Module directory in CMake of current module being processed: ``${ZEPHYR_CURRENT_MODULE_DIR}``
+- :file:`CMakeLists.txt` directory of current module directory: ``${ZEPHYR_CURRENT_CMAKE_DIR}``
```
```suggestion
- the current module's top level directory: ``${ZEPHYR_CURRENT_MODULE_DIR}``
- the current module's :file:`CMakeLists.txt` directory: ``${ZEPHYR_CURRENT_CMAKE_DIR}``
```

mbolivar-nordic commented on doc/guides/modules.rst:
```diff
@@ -75,6 +75,65 @@ module:
+- Module directory in CMake of current module being processed: ``${ZEPHYR_CURRENT_MODULE_DIR}``
+- :file:`CMakeLists.txt` directory of current module directory: ``${ZEPHYR_CURRENT_CMAKE_DIR}``
+
+This remove the need for a Zephyr module to know its own name during CMake
+processing, and the module can source additional CMake files using the current
+variable, for example:
```
```suggestion
This removes the need for a Zephyr module to know its own name during CMake
processing. The module can source additional CMake files using these ``CURRENT``
variables. For example:
```

mbolivar-nordic replied:
@tejlmand it occurs to me that this means no module can be named "current".

Worth enforcing in the build system?

tejlmand replied:
Good catch. Let's hope someone doesn't create a current measurement sw tool, and call it `current`.

tejlmand replied:
added extra commit to check for `current`: https://github.com/zephyrproject-rtos/zephyr/pull/27740/commits/8511f9fd9a854c1963a415a6a5e93681eaf0d79b

mbolivar-nordic replied:
> added extra commit to check for `current`: [8511f9f](https://github.com/zephyrproject-rtos/zephyr/commit/8511f9fd9a854c1963a415a6a5e93681eaf0d79b)

nice :+1:

mbolivar-nordic reviewed (changes requested):
I have a few cleanups I think would make the docs a bit clearer and then I am +1. Thanks again for taking the time to put together a proper fix.

dkalowsk commented:
It looks like @mbolivar-nordic has already suggested the same changes to the documentation I was in the process of creating, so I won't bother doubling up on those.  Appreciate the effort in cleaning and describing everything for the updated process and details.

tejlmand commented on doc/guides/modules.rst:
```diff
@@ -75,6 +75,65 @@ module:
+In Kconfig and CMake, the ``ZEPHYR_<module-name>_MODULE_DIR`` can be used to
+obtain the absolute path to the module directory.
+
+In CMake, the ``ZEPHYR_<module-name>_CMAKE_DIR`` can be used to obtain the
+absolute path to the :file:`CMakeLists.txt` that is included into CMake build
+system.
```
Not completely accurate. The  `ZEPHYR_<module-name>_CMAKE_DIR` is only dependent on CMakeLists.txt file.

tejlmand committed:
cmake: Zephyr module error on module name current

In CMake, ZEPHYR_CURRENT_MODULE_DIR is used during processing of module
CMakeLists.txt files.
If a module is named `current` then it's corresponding
`ZEPHYR_<module_name>_MODULE_DIR` would become ZEPHYR_CURRENT_MODULE_DIR
and thus colliding with internal Zephyr CMake variable.

Using module name `current` is therefore forbidden.
Print an error to users when a module named `current` is used in
Zephyr.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

mbolivar-nordic reviewed (approved):
Looks great! Thanks again.

mbolivar-nordic commented on cmake/zephyr_module.cmake:
```diff
@@ -59,6 +59,32 @@ if(WEST OR ZEPHYR_MODULES)
+
+      string(TOUPPER ${module_name} MODULE_NAME_UPPER)
+      if(NOT ${MODULE_NAME_UPPER} STREQUAL CURRENT)
+        set(ZEPHYR_${MODULE_NAME_UPPER}_MODULE_DIR ${cmake_path})
```
I rebased and retested the new version and it looks like a typo snuck in here. This should be:

```suggestion
        set(ZEPHYR_${MODULE_NAME_UPPER}_MODULE_DIR ${module_path})
```

tejlmand replied:
Good catch, thanks.

Wonder how that sneaked in, cause it is correct here: https://github.com/zephyrproject-rtos/zephyr/pull/27740/commits/67c09760e740f9872275592c99d23ffec7d52036
Guess i'll never know.

mbolivar-nordic reviewed (approved):
Working for me again, thanks for the fast update.

carlescufi commented:
Does this affect CI? https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/ci/check_compliance.py#L240

tejlmand commented:
> 
> 
> Does this affect CI? https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/ci/check_compliance.py#L240

No, this PR has no behavioral change on the build system, it only offers additional informational variables that can be used elsewhere in the build system. (and does some cleanup of existing build code)
